### ISSUE TITLE
DyGraph 动态图保存预测模型

### ISSUE THREAD

**Issue by yeyupiaoling** [PaddlePaddle/Paddle#18456]
# 环境
 - PaddlePaddle 1.5

# 问题
DyGraph 动态图没有提供保存预测模型的接口，跟Fluid的：save_inference_model
因为需要在 Paddle Mobile 中需求使用到预测模型，同时如果能够保存预测模型，在预测上就不需要再编写网络结构代码。

我尝试过使用 Executor 的方式训练 DyGraph 网络代码，但是在保存模型的时候出现问题
```
paddle.fluid.core_avx.EnforceNotMet: Invoke operator save error.
Python Callstacks: 
  File "C:\Users\1596\AppData\Local\Programs\Python\Python35\lib\site-packages\paddle\fluid\framework.py", line 1748, in append_op
    attrs=kwargs.get("attrs", None))
  File "C:\Users\1596\AppData\Local\Programs\Python\Python35\lib\site-packages\paddle\fluid\io.py", line 213, in save_vars
    'file_path': os.path.join(save_dirname, new_var.name)
  File "C:\Users\1596\AppData\Local\Programs\Python\Python35\lib\site-packages\paddle\fluid\io.py", line 191, in save_vars
    filename=filename)
  File "C:\Users\1596\AppData\Local\Programs\Python\Python35\lib\site-packages\paddle\fluid\io.py", line 509, in save_persistables
    filename=filename)
  File "C:\Users\1596\AppData\Local\Programs\Python\Python35\lib\site-packages\paddle\fluid\io.py", line 1087, in save_inference_model
    save_persistables(executor, save_dirname, main_program, params_filename)
  File "D:/yeyupiaoling/PyCharm/TestDyGraph/test7.py", line 222, in train
    executor=exe)
  File "D:/yeyupiaoling/PyCharm/TestDyGraph/test7.py", line 227, in <module>
    train()
C++ Callstacks: 
Cannot open infer_model\mnist/MNIST_0/FC_0.b_0 to write at [E:\release_cuda87\paddle\paddle/fluid/operators/save_op.h:82]
PaddlePaddle Call Stacks: 
Windows not support stack backtrace yet.
```

**Comment by heavengate** [comment]
看着是windows上不能识别infer_model\mnist/MNIST_0/FC_0.b_0中/的目录分隔符的问题，其中infer_model是save_dirname，但是new_var.name 为什么是mnist/MNIST_0/FC_0.b_0呢，麻烦贴一些关键代码确认一下

**Comment by yeyupiaoling** [comment]
代码如下：
```python
def train():
    epoch_num = 1
    BATCH_SIZE = 64
    # model, _ = fluid.dygraph.load_persistables('save_dir')
    mnist = MNIST("mnist")
    # mnist.load_dict(model)
    adam = fluid.optimizer.AdamOptimizer(learning_rate=0.001)
    train_reader = paddle.batch(paddle.dataset.mnist.train(), batch_size=BATCH_SIZE, drop_last=True)
    img = fluid.layers.data(name='pixel', shape=[1, 28, 28], dtype='float32')
    label = fluid.layers.data(name='label', shape=[1], dtype='int64')
    net, acc = mnist(img, label)
    loss = fluid.layers.cross_entropy(net, label)
    avg_loss = fluid.layers.mean(loss)
    adam.minimize(avg_loss)

    exe = fluid.Executor(fluid.CPUPlace())
    exe.run(fluid.default_startup_program())

    # 判断模型是否存在
    def if_exist(var):
        path = os.path.join('save_dir', var.name)
        exist = os.path.exists(path)
        if exist:
            print("Load model: %s" % path)
        return exist

    # 加载预训练模型
    fluid.io.load_vars(executor=exe, dirname='save_dir', predicate=if_exist)

    for epoch in range(epoch_num):
        for batch_id, data in enumerate(train_reader()):
            static_x_data = np.array([x[0].reshape(1, 28, 28) for x in data]).astype('float32')
            y_data = np.array([x[1] for x in data]).astype('int64').reshape([BATCH_SIZE, 1])

            fetch_list = [avg_loss.name, acc.name]
            out, a = exe.run(program=fluid.default_main_program(),
                             feed={"pixel": static_x_data,
                                   "label": y_data},
                             fetch_list=fetch_list)

            static_out = out[0]

            if batch_id % 100 == 0:
                print("epoch: {}, batch_id: {}, loss: {}, acc: {}".format(epoch, batch_id, static_out, a))

    fluid.io.save_inference_model(dirname='infer_model',
                                  feeded_var_names=[img.name],
                                  target_vars=[net],
                                  executor=exe)
```

**Comment by yeyupiaoling** [comment]
@heavengate 在Windows下，这种路径是没有问题的。

**Comment by heavengate** [comment]
看你代码网络构建在，net, acc = mnist(img, label)这个里面，是否是这个里面定义的变量名是mnist/MNIST_0/FC_0.b_0这种形式windows无法识别，可改成mnist\MNIST_0\FC_0.b_0这种形式

**Comment by yeyupiaoling** [comment]
@heavengate 全部代码给你看看吧
```python
from __future__ import print_function

import os

import numpy as np
import paddle
import paddle.fluid as fluid
from PIL import Image


class SimpleImgConvPool(fluid.dygraph.Layer):
    def __init__(self,
                 name_scope,
                 num_filters,
                 filter_size,
                 pool_size,
                 pool_stride,
                 pool_padding=0,
                 pool_type='max',
                 global_pooling=False,
                 conv_stride=1,
                 conv_padding=0,
                 conv_dilation=1,
                 conv_groups=1,
                 act=None,
                 use_cudnn=False,
                 param_attr=None,
                 bias_attr=None):
        super(SimpleImgConvPool, self).__init__(name_scope)

        self._conv2d = fluid.dygraph.Conv2D(
            self.full_name(),
            num_filters=num_filters,
            filter_size=filter_size,
            stride=conv_stride,
            padding=conv_padding,
            dilation=conv_dilation,
            groups=conv_groups,
            param_attr=None,
            bias_attr=None,
            use_cudnn=use_cudnn)

        self._pool2d = fluid.dygraph.Pool2D(
            self.full_name(),
            pool_size=pool_size,
            pool_type=pool_type,
            pool_stride=pool_stride,
            pool_padding=pool_padding,
            global_pooling=global_pooling,
            use_cudnn=use_cudnn)

    def forward(self, inputs):
        x = self._conv2d(inputs)
        x = self._pool2d(x)
        return x


class MNIST(fluid.dygraph.Layer):
    def __init__(self, name_scope):
        super(MNIST, self).__init__(name_scope)

        self._simple_img_conv_pool_1 = SimpleImgConvPool(self.full_name(), 20, 5, 2, 2, act="relu")
        self._simple_img_conv_pool_2 = SimpleImgConvPool(self.full_name(), 50, 5, 2, 2, act="relu")

        pool_2_shape = 50 * 4 * 4
        SIZE = 10
        scale = (2.0 / (pool_2_shape ** 2 * SIZE)) ** 0.5
        self._fc = fluid.dygraph.FC(self.full_name(),
                                    size=10,
                                    param_attr=fluid.param_attr.ParamAttr(
                                        initializer=fluid.initializer.NormalInitializer(loc=0.0, scale=scale)),
                                    act="softmax")

    def forward(self, inputs, label=None):
        x = self._simple_img_conv_pool_1(inputs)
        x = self._simple_img_conv_pool_2(x)
        x = self._fc(x)
        if label is not None:
            acc = fluid.layers.accuracy(input=x, label=label)
            return x, acc
        else:
            return x


def test_train(reader, model, batch_size):
    acc_set = []
    avg_loss_set = []
    for batch_id, data in enumerate(reader()):
        dy_x_data = np.array([x[0].reshape(1, 28, 28) for x in data]).astype('float32')
        y_data = np.array([x[1] for x in data]).astype('int64').reshape(batch_size, 1)

        img = fluid.dygraph.base.to_variable(dy_x_data)
        label = fluid.dygraph.base.to_variable(y_data)
        label.stop_gradient = True
        prediction, acc = model(img, label)
        loss = fluid.layers.cross_entropy(input=prediction, label=label)
        avg_loss = fluid.layers.mean(loss)
        acc_set.append(float(acc.numpy()))
        avg_loss_set.append(float(avg_loss.numpy()))

        # get test acc and loss
    acc_val_mean = np.array(acc_set).mean()
    avg_loss_val_mean = np.array(avg_loss_set).mean()

    return avg_loss_val_mean, acc_val_mean


def train_mnist():
    epoch_num = 1
    BATCH_SIZE = 64
    with fluid.dygraph.guard(place=fluid.CPUPlace()):

        mnist = MNIST("mnist")
        model, _ = fluid.dygraph.load_persistables("save_dir")
        mnist.load_dict(model)
        adam = fluid.optimizer.AdamOptimizer(learning_rate=0.001)
        train_reader = paddle.batch(paddle.dataset.mnist.train(), batch_size=BATCH_SIZE, drop_last=True)
        test_reader = paddle.batch(paddle.dataset.mnist.test(), batch_size=BATCH_SIZE, drop_last=True)

        for epoch in range(epoch_num):
            for batch_id, data in enumerate(train_reader()):
                dy_x_data = np.array([x[0].reshape(1, 28, 28) for x in data]).astype('float32')
                y_data = np.array([x[1] for x in data]).astype('int64').reshape(BATCH_SIZE, 1)

                img = fluid.dygraph.base.to_variable(dy_x_data)
                label = fluid.dygraph.base.to_variable(y_data)
                label.stop_gradient = True

                cost, acc = mnist(img, label)

                loss = fluid.layers.cross_entropy(cost, label)
                avg_loss = fluid.layers.mean(loss)
                avg_loss.backward()
                adam.minimize(avg_loss)
                # save checkpoint
                mnist.clear_gradients()
                if batch_id % 100 == 0:
                    print("Loss at epoch {} step {}: {:}".format(epoch, batch_id, avg_loss.numpy()))

            mnist.eval()
            test_cost, test_acc = test_train(test_reader, mnist, BATCH_SIZE)
            mnist.train()
            print("Loss at epoch {} , Test avg_loss is: {}, acc is: {}".format(epoch, test_cost, test_acc))

        fluid.dygraph.save_persistables(mnist.state_dict(), "save_dir")
        print("checkpoint saved")

    with fluid.dygraph.guard(place=fluid.CPUPlace()):

        mnist_infer = MNIST("mnist")
        # load checkpoint
        model, _ = fluid.dygraph.load_persistables("save_dir")
        mnist_infer.load_dict(model)
        print("checkpoint loaded")

        # start evaluate mode
        mnist_infer.eval()

        def load_image(file):
            im = Image.open(file).convert('L')
            im = im.resize((28, 28), Image.ANTIALIAS)
            im = np.array(im).reshape(1, 1, 28, 28).astype(np.float32)
            im = im / 255.0 * 2.0 - 1.0
            return im

        tensor_img = load_image('infer_3.png')

        results = mnist_infer(fluid.dygraph.base.to_variable(tensor_img))
        lab = np.argsort(results.numpy())
        print("Inference result of image/infer_3.png is: %d" % lab[0][-1])
        print(results.numpy()[0][lab[0][-1]])


def train():
    epoch_num = 1
    BATCH_SIZE = 64
    mnist = MNIST("mnist")
    # model, _ = fluid.dygraph.load_persistables('save_dir')
    # mnist.load_dict(model)
    adam = fluid.optimizer.AdamOptimizer(learning_rate=0.001)
    train_reader = paddle.batch(paddle.dataset.mnist.train(), batch_size=BATCH_SIZE, drop_last=True)
    img = fluid.layers.data(name='pixel', shape=[1, 28, 28], dtype='float32')
    label = fluid.layers.data(name='label', shape=[1], dtype='int64')
    net, acc = mnist(img, label)
    loss = fluid.layers.cross_entropy(net, label)
    avg_loss = fluid.layers.mean(loss)
    adam.minimize(avg_loss)

    exe = fluid.Executor(fluid.CPUPlace())
    exe.run(fluid.default_startup_program())

    # 判断模型是否存在
    def if_exist(var):
        path = os.path.join('save_dir', var.name)
        exist = os.path.exists(path)
        if exist:
            print("Load model: %s" % path)
        return exist

    # 加载预训练模型
    fluid.io.load_vars(executor=exe, dirname='save_dir', predicate=if_exist)

    for epoch in range(epoch_num):
        for batch_id, data in enumerate(train_reader()):
            static_x_data = np.array([x[0].reshape(1, 28, 28) for x in data]).astype('float32')
            y_data = np.array([x[1] for x in data]).astype('int64').reshape([BATCH_SIZE, 1])

            fetch_list = [avg_loss.name, acc.name]
            out, a = exe.run(program=fluid.default_main_program(),
                             feed={"pixel": static_x_data,
                                   "label": y_data},
                             fetch_list=fetch_list)

            static_out = out[0]

            if batch_id % 100 == 0:
                print("epoch: {}, batch_id: {}, loss: {}, acc: {}".format(epoch, batch_id, static_out, a))

    fluid.io.save_inference_model(dirname='infer_model',
                                  feeded_var_names=[img.name],
                                  target_vars=[net],
                                  executor=exe)


if __name__ == '__main__':
    # 动态图
    # train_mnist()
    # 静态图
    train()

```

**Comment by heavengate** [comment]
问题原因如下：
SimpleImgConvPool/MNIST这些网络结构都继承自fluid.dygraph.Layer，参数名统一用的是fluid.dygraph.Layer的full_name方法，该方法定义如下
https://github.com/PaddlePaddle/Paddle/blob/a873fa84ceca411a5a776ff8ae303f8be24df95a/python/paddle/fluid/dygraph/layers.py#L43
是采用的linux下的目录分隔符/，该分隔符不兼容windows下

规避方法：windows上在所有full_name()后加 .replace('/', '\\')改成windows风格的目录分隔符

该问题我们已记录，并将立即修复

**Comment by yeyupiaoling** [comment]
@heavengate 嗯嗯，解决了，谢谢。
修改的代码片段如下：
```python
class MNIST(fluid.dygraph.Layer):
    def __init__(self, name_scope):
        super(MNIST, self).__init__(name_scope)

        self._simple_img_conv_pool_1 = SimpleImgConvPool(self.full_name().replace('/', '\\'), 20, 5, 2, 2, act="relu")
        self._simple_img_conv_pool_2 = SimpleImgConvPool(self.full_name().replace('/', '\\'), 50, 5, 2, 2, act="relu")

        pool_2_shape = 50 * 4 * 4
        SIZE = 10
        scale = (2.0 / (pool_2_shape ** 2 * SIZE)) ** 0.5
        self._fc = fluid.dygraph.FC(self.full_name().replace('/', '\\'),
                                    size=10,
                                    param_attr=fluid.param_attr.ParamAttr(
                                        initializer=fluid.initializer.NormalInitializer(loc=0.0, scale=scale)),
                                    act="softmax")

    def forward(self, inputs, label=None):
        x = self._simple_img_conv_pool_1(inputs)
        x = self._simple_img_conv_pool_2(x)
        x = self._fc(x)
        if label is not None:
            acc = fluid.layers.accuracy(input=x, label=label)
            return x, acc
        else:
            return x
```

**Comment by yeyupiaoling** [comment]
@heavengate 顺便问一下，为什么不在DyGraph 动态图直接提供保存预测模型接口？

**Comment by heavengate** [comment]
> @heavengate 顺便问一下，为什么不在DyGraph 动态图直接提供保存预测模型接口？

动态图目前release是初步版本，还有很多特性在进一步开发完善

**Comment by yeyupiaoling** [comment]
@heavengate 哦哦，那再问一下，DyGraph 动态图回事之后主推的版本吗？
会不会像当初的Fluid那样，抛弃了V2，主推Fluid。

**Comment by heavengate** [comment]
> @heavengate 哦哦，那再问一下，DyGraph 动态图回事之后主推的版本吗？
> 会不会像当初的Fluid那样，抛弃了V2，主推Fluid。

动态图是Fluid的一大特性，是Fluid框架的一部分

**Comment by yeyupiaoling** [comment]
@heavengate 是吗。我的理解它们是两个完全不同的版本，执行方式都不一样，就像当初的V2和Fluid一样，完全两个不同的版本。

**Comment by heavengate** [comment]
> @heavengate 是吗。我的理解它们是两个完全不同的版本，执行方式都不一样，就像当初的V2和Fluid一样，完全两个不同的版本。

后面会调整，不会是不同版本的

**Comment by yeyupiaoling** [comment]
@heavengate 好的，谢谢回答。
